Process valid lines where possible

domleb opened the pull request:
The current implementation will drop all points in a batch if any of them are invalid.  Instead, process all the valid lines and drop only the invalid ones.  Keep the existing behavior of returning StatusBadRequest if any lines are invalid and return a json response with success / failure counts for the batch.

domleb committed:
Process valid lines where possible

The current implementation will drop all points in a batch if any of them are invalid.  Instead, process all the valid lines and drop only the invalid ones.  Keep the existing behavior of returning StatusBadRequest if any lines are invalid and return a json response with success / failure counts for the batch.

wadawe commented on receiver/influxdbreceiver/receiver.go:
```diff
@@ -201,6 +212,23 @@ func (r *metricsReceiver) handleWrite(w http.ResponseWriter, req *http.Request)
+		for _, count := range failureReason {
+			totalErrors += count
+		}
+		w.WriteHeader(http.StatusBadRequest)
+		w.Header().Set("Content-Type", "application/json")
```
Setting headers should come before `.WriteHeader`.
```suggestion
		w.Header().Set("Content-Type", "application/json")
		w.WriteHeader(http.StatusBadRequest)
```

wadawe commented on receiver/influxdbreceiver/receiver.go:
```diff
@@ -201,6 +212,23 @@ func (r *metricsReceiver) handleWrite(w http.ResponseWriter, req *http.Request)
+		response := map[string]any{
+			"total_success":  line - totalErrors,
+			"total_failure":  totalErrors,
+			"failure_reason": failureReason,
```
I'd also maybe be a nit-picking nancy and suggest this to be `failure_reasons`?
```suggestion
			"failure_reasons": failureReason,
```

wadawe reviewed (commented):
Only some minor changes

domleb committed:
General improvements

- Match the behavior of InfluxDBs partial writes so that the entire batch isn't dropped if there is a single line with a parsing error.
- Add config toggle for partial writes.  Default to false so behavior is the same for existing users of the receiver.
- Add config to limit tracking of errors to improve performance (InfluxDB doesn't do this, it always returns all errors plus it returns the full line for every error - instead we just return the line number, measurement name and error message).  Default to 0 which still returns an error message saying 'partial write' as per InfluxDB.

domleb commented:
Improved implementation based on how InfluxDB handles it by doing partial writes
